File: src/main/java/ru/timeconqueror/timecore/common/capability/property/serializer/NullPropertySerializer.java
package ru.timeconqueror.timecore.common.capability.property.serializer;

import lombok.RequiredArgsConstructor;
import net.minecraft.nbt.CompoundTag;

@RequiredArgsConstructor
public class NullPropertySerializer<T> implements IPropertySerializer<T> {
    private final IPropertySerializer<T> prop;

    @Override
    public void serialize(String name, T value, CompoundTag nbt) {
        CompoundTag comp = new CompoundTag();
        if (value != null) {
            prop.serialize("value", value, comp);
        }
        nbt.put(name, comp);
    }

    @Override
    public T deserialize(String name, CompoundTag nbt) {
        CompoundTag comp = nbt.getCompound(name);
        if (comp.contains("value")) {
            return prop.deserialize("value", comp);
        }
        return null;
    }
}

File: src/main/java/ru/timeconqueror/timecore/common/capability/property/serializer/NullPropertySerializer.kt
package ru.timeconqueror.timecore.common.capability.property.serializer

import net.minecraft.nbt.CompoundTag


open class NullPropertySerializer<T>(val prop: IPropertySerializer<T>) : IPropertySerializer<T?> {

    override fun serialize(name: String, value: T?, nbt: CompoundTag) {
        val comp = CompoundTag()
        if (value != null) {
            prop.serialize("value", value, comp)
        }
        nbt.put(name, comp)
    }

    override fun deserialize(name: String, nbt: CompoundTag): T? {
        val comp = nbt.getCompound(name)
        if (comp.contains("value")) {
            return prop.deserialize("value", comp)
        }
        return null
    }


}